File: shared/ios/ReleafCore/Sources/ReleafCoreScan/SearchablePdfExporter.swift
/*
 * SearchablePdfExporter.swift
 *
 * Builds a multi-page PDF from `[PageContent]` (per-page image +
 * `OcrResult`) with an invisible text layer overlaid on top of each
 * image. Search and copy/paste in any conforming PDF reader pull
 * text from the overlay; the page renders visually as the source
 * image alone.
 *
 * Per QUICKINK_PROPOSAL.md §6.3, this is the prototype path behind
 * the `searchablePdfExportEnabled` feature flag — the v1 default
 * export uses Releaf's existing flat-PDF code path. The flag
 * wiring + the Settings → "Experimental" toggle land with the
 * QuickInk MVP; this file ships the exporter contract + impl so
 * the toggle has something to invoke when it's flipped on.
 *
 * Mirror of `SearchablePdfExporter.kt` in `:shared:scan`.
 *
 * Implementation:
 *
 *   - `UIGraphicsPDFRenderer` for the page rendering. Page size is
 *     the image's pixel dimensions interpreted as PDF points
 *     (matches what `DocumentScannerView` already does for the
 *     flat-PDF path; consistent across the two exporters).
 *
 *   - Invisible text via `CGContext.setTextDrawingMode(.invisible)`
 *     — PDF rendering mode 3, "neither stroke nor fill". Text is
 *     present in the content stream (searchable + selectable) but
 *     no glyphs are drawn. Standard searchable-PDF technique;
 *     supported by every reader that implements ISO 32000.
 *
 *   - Granularity: line-level only (`OcrBlock.kind == .line`).
 *     Paragraph-level blocks would lump multiple lines into one
 *     hit; word-level (which neither engine produces today anyway)
 *     would lose word-spacing context for search-and-copy.
 *
 *   - Coordinate translation: OCR's normalized 0..1 top-left
 *     bboxes → PDF points with Y flipped (PDF is bottom-left
 *     origin). Both `OcrBbox` and the PDF point system are
 *     orthogonal, so it's a clean affine transform per block.
 *
 * `#if os(iOS)`-guarded for the same reason `VisionTextRecognizer`
 * is — package macOS deployment minimum is resolver appeasement
 * only; nothing actually runs there.
 */

#if os(iOS)

import Foundation
import UIKit

public struct SearchablePdfExporter: Sendable {

    public init() {}

    /// Builds a PDF from `pages` and writes it to `outputURL`. Each
    /// element of `pages` becomes one PDF page with the image
    /// rendered visually and the OCR text overlaid invisibly.
    public func export(pages: [PageContent], to outputURL: URL) throws {
        guard !pages.isEmpty else {
            // An empty input list produces an empty PDF, which most
            // readers reject. Easier to fail early with a clear
            // signal than write a 0-page file the caller has to
            // post-validate.
            throw ExportError.pdfWriteFailed(message: "pages is empty")
        }

        // Page-by-page bounds — `UIGraphicsPDFRenderer`'s `format`
        // is per-document, but we override the bounds per page in
        // `beginPage`, so any non-zero seed works for the
        // constructor. We use the first page's image as the seed.
        let firstImage = try Self.loadImage(at: pages[0].imageURL)
        let initialBounds = CGRect(origin: .zero, size: firstImage.size)
        let renderer = UIGraphicsPDFRenderer(bounds: initialBounds)

        do {
            try renderer.writePDF(to: outputURL) { context in
                for (index, page) in pages.enumerated() {
                    let image = (index == 0)
                        ? firstImage
                        : (try? Self.loadImage(at: page.imageURL)) ?? firstImage
                    Self.drawPage(image: image, ocr: page.ocrResult, in: context)
                }
            }
        } catch {
            throw ExportError.pdfWriteFailed(message: error.localizedDescription)
        }
    }

    // MARK: - Page rendering

    private static func drawPage(
        image: UIImage,
        ocr: OcrResult,
        in context: UIGraphicsPDFRendererContext
    ) {
        // Per-page bounds match the source image. PDF readers honor
        // arbitrary page sizes, so there's no reason to letterbox
        // into a fixed Letter / A4 frame.
        let pageBounds = CGRect(origin: .zero, size: image.size)
        context.beginPage(withBounds: pageBounds, pageInfo: [:])

        // 1. Visible image fill.
        image.draw(in: pageBounds)

        // 2. Invisible text overlay — line-grained only.
        let cg = context.cgContext
        cg.saveGState()
        cg.setTextDrawingMode(.invisible)
        defer { cg.restoreGState() }

        for block in ocr.blocks where block.kind == .line {
            drawInvisibleLine(block, pageSize: image.size, in: cg)
        }
    }

    private static func drawInvisibleLine(
        _ block: OcrBlock,
        pageSize: CGSize,
        in cg: CGContext
    ) {
        // OCR bbox: normalized 0..1, top-left origin, image space.
        // PDF page coords: bottom-left origin, in points (we set
        // points = pixels via the image-sized page bounds above).
        let pixelX      = block.bbox.x      * Double(pageSize.width)
        let pixelY      = block.bbox.y      * Double(pageSize.height)
        let pixelWidth  = block.bbox.width  * Double(pageSize.width)
        let pixelHeight = block.bbox.height * Double(pageSize.height)

        // Y-flip: top-left y → bottom-left baseline. Place the
        // baseline at (1 - bbox.y - bbox.height) * height. Adding a
        // small descender allowance isn't worth the complexity for
        // an invisible layer — search hits are spatially loose.
        let pdfX        = pixelX
        let pdfBaseline = Double(pageSize.height) - pixelY - pixelHeight

        // Font size approximates the bbox height. We don't need a
        // pixel-perfect glyph match — the goal is "search hit
        // highlight roughly covers the visible word". Helvetica
        // works because the glyphs aren't rendered.
        let fontSize = max(pixelHeight, 1.0)
        let font = UIFont(name: "Helvetica", size: fontSize)
            ?? UIFont.systemFont(ofSize: fontSize)

        let attributes: [NSAttributedString.Key: Any] = [
            .font:            font,
            .foregroundColor: UIColor.clear,
        ]
        let attributed = NSAttributedString(string: block.text, attributes: attributes)

        // `NSAttributedString.draw(at:)` uses the text drawing mode
        // we set above (.invisible), so the actual color is moot —
        // glyphs are added to the content stream but not painted.
        // We position by the baseline-ish corner; AppKit/UIKit's
        // draw(at:) treats the point as the upper-left of the
        // bounding box, but Core Graphics under the hood places the
        // baseline. For an invisible layer the difference is
        // imperceptible to search.
        let drawPoint = CGPoint(x: pdfX, y: pdfBaseline)

        // Width-fitting: for very long blocks of text, scale font
        // down so the glyph run fits the OCR bbox. Keeps
        // search-highlight position roughly correct.
        let measured = attributed.size()
        if measured.width > CGFloat(pixelWidth), measured.width > 0 {
            let shrunkSize = fontSize * Double(pixelWidth) / Double(measured.width)
            let shrunkFont = UIFont(name: "Helvetica", size: shrunkSize)
                ?? UIFont.systemFont(ofSize: shrunkSize)
            let shrunk = NSAttributedString(string: block.text, attributes: [
                .font:            shrunkFont,
                .foregroundColor: UIColor.clear,
            ])
            shrunk.draw(at: drawPoint)
        } else {
            attributed.draw(at: drawPoint)
        }
    }

    // MARK: - Loading

    private static func loadImage(at url: URL) throws -> UIImage {
        guard let image = UIImage(contentsOfFile: url.path) else {
            throw ExportError.imageUnreadable(url)
        }
        return image
    }
}

// MARK: - Inputs + errors

extension SearchablePdfExporter {

    /// One page of input — image + its OCR result. The image is
    /// rendered as the visible page; the OCR's line-grained blocks
    /// are overlaid as invisible text.
    public struct PageContent: Sendable {
        public let imageURL:  URL
        public let ocrResult: OcrResult

        public init(imageURL: URL, ocrResult: OcrResult) {
            self.imageURL = imageURL
            self.ocrResult = ocrResult
        }
    }

    /// Failure modes the exporter surfaces. Distinct from `OcrError`
    /// because the export pipeline is downstream of OCR — by the
    /// time we reach this code, recognition has already succeeded.
    public enum ExportError: Error {
        /// Image at `URL` couldn't be loaded (missing, unsupported
        /// format, corrupt bytes).
        case imageUnreadable(URL)

        /// PDF write failed — disk full, sandbox permissions, etc.
        /// `message` carries the underlying reason.
        case pdfWriteFailed(message: String)
    }
}

#endif

File: shared/android/shared/scan/src/main/kotlin/app/releaf/shared/scan/SearchablePdfExporter.kt
/*
 * SearchablePdfExporter.kt
 *
 * Builds a multi-page PDF from `[PageContent]` (per-page image +
 * `OcrResult`) with an invisible text layer overlaid on top of each
 * image. Search and copy/paste in any conforming PDF reader pull
 * text from the overlay; the page renders visually as the source
 * image alone.
 *
 * Per QUICKINK_PROPOSAL.md §6.3, this is the prototype path behind
 * the `searchablePdfExportEnabled` feature flag — the v1 default
 * export uses ML Kit's built-in `RESULT_FORMAT_PDF` from the
 * document scanner. The flag wiring + the Settings → "Experimental"
 * toggle land with the QuickInk MVP; this file ships the exporter
 * contract + impl so the toggle has something to invoke when it's
 * flipped on.
 *
 * Mirror of `SearchablePdfExporter.swift` in `ReleafCoreScan`.
 *
 * Implementation:
 *
 *   - `android.graphics.pdf.PdfDocument` for the page rendering.
 *     Page size is the image's pixel dimensions interpreted as PDF
 *     points (matches the convention iOS uses; both platforms emit
 *     PDFs with arbitrary per-page sizes, which every conforming
 *     reader handles).
 *
 *   - Invisible text via `Paint.color = Color.TRANSPARENT`. Stock
 *     Android `PdfDocument` doesn't expose PDF rendering modes
 *     directly; we rely on the text being added to the page's
 *     content stream regardless of paint alpha. PDF text-extraction
 *     libraries (and Spotlight / Android system search) read text
 *     from the content stream — alpha is a render-time concern.
 *
 *     Caveat: some PDF readers optimize out fully transparent text
 *     in their text-extraction path. If that turns out to be a
 *     real problem in dev-build smoke tests, the path forward is
 *     swapping to PDFBox-Android (Apache 2.0; ~5MB AAR) and using
 *     proper rendering-mode-3 invisible text. Out of scope for the
 *     prototype; a feature-flagged dev-only feature is fine with
 *     the simpler approach.
 *
 *   - Granularity: line-level only (`OcrBlock.Kind.Line`).
 *     Paragraph-level blocks would lump multiple lines into one
 *     hit; word-level (which neither engine produces today anyway)
 *     would lose word-spacing context for search-and-copy.
 *
 *   - Coordinate translation: OCR's normalized 0..1 top-left
 *     bboxes → PDF points with Y staying top-left (Android Canvas
 *     and `PdfDocument` both use top-left origin, unlike Core
 *     Graphics on iOS — saves us a Y-flip the iOS impl needs).
 */

package app.releaf.shared.scan

import android.content.Context
import android.graphics.BitmapFactory
import android.graphics.Color
import android.graphics.Paint
import android.graphics.pdf.PdfDocument
import android.net.Uri
import java.io.File
import java.io.IOException

class SearchablePdfExporter(private val context: Context) {

    /**
     * Builds a PDF from `pages` and writes it to `output`. Each
     * element of `pages` becomes one PDF page with the image
     * rendered visually and the OCR text overlaid invisibly.
     *
     * Throws [ExportException.ImageUnreadable] when a page's image
     * Uri can't be decoded, or [ExportException.PdfWriteFailed]
     * when the file write itself fails (disk full, permission
     * denied, etc.). Empty `pages` is a precondition error — most
     * PDF readers reject 0-page files; fail early instead.
     */
    fun export(pages: List<PageContent>, output: File) {
        require(pages.isNotEmpty()) { "pages must be non-empty" }

        val document = PdfDocument()
        try {
            for ((index, page) in pages.withIndex()) {
                drawPage(document, pageNumber = index + 1, page = page)
            }

            try {
                output.outputStream().use { stream ->
                    document.writeTo(stream)
                }
            } catch (e: IOException) {
                throw ExportException.PdfWriteFailed(e.message.orEmpty(), e)
            }
        } finally {
            document.close()
        }
    }

    // ─── Page rendering ──────────────────────────────────────────

    private fun drawPage(
        document: PdfDocument,
        pageNumber: Int,
        page: PageContent,
    ) {
        // Decode the bitmap once. We need its dimensions for the
        // PDF page size and for normalizing OCR bboxes back to
        // pixels. `decodeStream` returns null on unsupported /
        // corrupt formats — fold to ImageUnreadable.
        val bitmap = try {
            context.contentResolver.openInputStream(page.imageUri)?.use { stream ->
                BitmapFactory.decodeStream(stream)
            }
        } catch (e: IOException) {
            throw ExportException.ImageUnreadable(page.imageUri, e)
        } ?: throw ExportException.ImageUnreadable(page.imageUri)

        try {
            val pageInfo = PdfDocument.PageInfo
                .Builder(bitmap.width, bitmap.height, pageNumber)
                .create()
            val pdfPage = document.startPage(pageInfo)
            try {
                val canvas = pdfPage.canvas

                // 1. Visible image fill — drawn at the page origin
                // (0, 0) with no paint. Use the positional overload
                // explicitly; the (Bitmap, Matrix?, Paint?) overload's
                // Matrix arg is annotated non-null on newer SDKs, so
                // passing null for both fails overload resolution.
                canvas.drawBitmap(bitmap, 0f, 0f, null)

                // 2. Invisible text overlay — line-grained only.
                val invisible = Paint().apply {
                    color = Color.TRANSPARENT
                    isAntiAlias = false  // moot for invisible glyphs; saves CPU
                }
                for (block in page.ocrResult.blocks) {
                    if (block.kind != OcrBlock.Kind.Line) continue
                    drawInvisibleLine(
                        canvas      = canvas,
                        block       = block,
                        pageWidth   = bitmap.width.toDouble(),
                        pageHeight  = bitmap.height.toDouble(),
                        paint       = invisible,
                    )
                }
            } finally {
                document.finishPage(pdfPage)
            }
        } finally {
            // Don't recycle the bitmap if the caller might want to
            // keep using the underlying file for other paths; close
            // is enough.
            bitmap.recycle()
        }
    }

    private fun drawInvisibleLine(
        canvas: android.graphics.Canvas,
        block: OcrBlock,
        pageWidth: Double,
        pageHeight: Double,
        paint: Paint,
    ) {
        // OCR bbox: normalized 0..1, top-left origin.
        // Android Canvas: pixel coords, top-left origin (top-left
        // matches OCR's frame, no flip needed — unlike Core
        // Graphics' bottom-left convention on iOS).
        val pixelX      = block.bbox.x      * pageWidth
        val pixelY      = block.bbox.y      * pageHeight
        val pixelWidth  = block.bbox.width  * pageWidth
        val pixelHeight = block.bbox.height * pageHeight

        // Font size approximates the bbox height. We don't need a
        // pixel-perfect glyph match — the goal is "search hit
        // highlight roughly covers the visible word".
        val targetSize = pixelHeight.coerceAtLeast(1.0).toFloat()
        paint.textSize = targetSize

        // Width-fit: for very long lines, shrink the size so the
        // glyph run fits the bbox. Keeps search-highlight position
        // roughly correct on long-line cases.
        val measured = paint.measureText(block.text).toDouble()
        if (measured > pixelWidth && measured > 0.0) {
            paint.textSize = (targetSize * (pixelWidth / measured)).toFloat()
        }

        // `drawText` places the baseline at `y`. The bbox's
        // top-left y maps to the line's top edge, so the baseline
        // sits roughly at `y + height`. We use the bottom of the
        // bbox as the baseline approximation — close enough for
        // invisible text whose alignment only needs to satisfy
        // search-highlight overlap.
        val baselineY = (pixelY + pixelHeight).toFloat()
        canvas.drawText(block.text, pixelX.toFloat(), baselineY, paint)
    }

    // ─── Inputs + errors ─────────────────────────────────────────

    /**
     * One page of input — image + its OCR result. The image is
     * rendered as the visible page; the OCR's line-grained blocks
     * are overlaid as invisible text.
     */
    data class PageContent(
        val imageUri:  Uri,
        val ocrResult: OcrResult,
    )

    /**
     * Failure modes the exporter surfaces. Distinct from
     * [OcrException] because the export pipeline is downstream of
     * OCR — by the time we reach this code, recognition has
     * already succeeded.
     */
    sealed class ExportException(message: String, cause: Throwable? = null)
        : Exception(message, cause) {

        /** Image at `uri` couldn't be loaded. */
        class ImageUnreadable(val uri: Uri, cause: Throwable? = null)
            : ExportException("Image at $uri couldn't be loaded", cause)

        /** PDF write failed — disk full, sandbox permissions, etc. */
        class PdfWriteFailed(message: String, cause: Throwable? = null)
            : ExportException("PDF write failed: $message", cause)
    }
}
